tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java

```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
-                    batchRows.clear();
-                }
-            }
-            if (preparedStatement != null) {
-                executeBatchWithVerification(preparedStatement, rowsInserted - batchRows.size(), batchRows);
             }
+            executeBatchWithVerification(preparedStatement, batch);
         }
-        catch (SQLException e) {
-            throw new RuntimeException(e);
+
+        private void executeBatchWithVerification(PreparedStatement preparedStatement, List<List<Object>> batch)
+                throws SQLException
+        {
+            if (batch.size() == 0) {
+                return;
+            }
+            int[] insertCounts = preparedStatement.executeBatch();
+            for (int rowIndex = 0; rowIndex < insertCounts.length; ++rowIndex) {
+                if (insertCounts[rowIndex] != 1) {
+                    throw new RuntimeException("could not insert values=" + batch.get(rowIndex));
+                }
+            }
         }
-        finally {
+
+        @Override
+        public void close()
+                throws SQLException
+        {
             if (preparedStatement != null) {
-                try {
-                    preparedStatement.close();
-                }
-                catch (SQLException e) {
-                    throw new RuntimeException(e);
-                }
+                preparedStatement.close();
             }
         }
     }
 
-    private void executeBatchWithVerification(PreparedStatement preparedStatement, int baseBatchRowIndex, List<List<Object>> batchRows)
-            throws SQLException
+    private static class InsertLoader
+            implements Loader
     {
-        checkNotNull(preparedStatement);
-        if (batchRows.size() == 0) {
-            return;
+        private final QueryExecutor queryExecutor;
+        private final String insertSql;
+        private final List<JDBCType> columnTypes;
+
+        public InsertLoader(QueryExecutor queryExecutor, String insertSql, String tableName)
+                throws SQLException
+        {
+            this.queryExecutor = requireNonNull(queryExecutor, "queryExecutor is null");
+            this.insertSql = requireNonNull(insertSql, "insertSql is null");
+            columnTypes = queryExecutor.executeQuery("SELECT * FROM " + tableName + " LIMIT 1").getColumnTypes();
         }
-        int[] insertCounts = preparedStatement.executeBatch();
-        for (int rowIndex = 0; rowIndex < insertCounts.length; ++rowIndex) {
-            if (insertCounts[rowIndex] != 1) {
-                throw new RuntimeException("could not insert values for row " + (baseBatchRowIndex + rowIndex) + "; values=" + batchRows.get(rowIndex));
+
+        @Override
+        public void load(List<List<Object>> batch)
+                throws SQLException
+        {
+            for (List<Object> row : batch) {
+                queryExecutor.executeQuery(insertSql, asQueryParams(row).toArray(new QueryExecutor.QueryParam[0]));
```

Comment:
Wouldn't `QueryExecutor` try using prepared statements when `QueryParams` are given?
`JdbcQueryExecutor` seems base on `PreparedStatement` in `executeQueryWithParams` method.

Why don't use just values without `QueryParams`? It would also save us issuing additional query (here: https://github.com/prestodb/tempto/pull/150/commits/8a8cc07040617168e94f94bffac7d407aac8bba8#diff-33d694596851b3e9776aefc8165805ffR338).